go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -1030,3 +1032,36 @@ func (vr *vreplicator) newClientConnection(ctx context.Context) (*vdbClient, err
 	}
 	return dbClient, nil
 }
+
+func (vr *vreplicator) readExistingRowsCopiedAndUpdateSelfIfNeeded() {
+	if vr.stats.CopyRowCount.Get() == 0 {
+		rowsCopiedExisting, err := vr.readExistingRowsCopied(vr.id)
+		if err != nil {
+			log.Warningf("failed to read existing rows copied: %v", err)
+		} else if rowsCopiedExisting != 0 {
+			log.Infof("resuming vreplication from another tablet, setting rows copied counter to %v", rowsCopiedExisting)
+			vr.stats.CopyRowCount.Set(rowsCopiedExisting)
+		}
+	}
+}
+
+func (vr *vreplicator) readExistingRowsCopied(uid int32) (int64, error) {
+	query, err := sqlparser.ParseAndBind(`SELECT rows_copied FROM _vt.vreplication WHERE id=%a`,
+		sqltypes.Int32BindVariable(uid),
+	)
+	if err != nil {
+		return 0, err
+	}
+	r, err := vr.dbClient.Execute(query)
+	if err != nil {
+		return 0, err
+	}
+	if len(r.Rows) == 0 {
+		return 0, nil
+	}
+	row := r.Named().Row()
+	if row == nil {
+		return 0, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "Cannot find unique workflow for UUID: %+v", uid)
+	}
+	return row.AsInt64("rows_copied", 0), nil
+}
```

Comment:
```suggestion
func (vr *vreplicator) readExistingRowsCopied(id int32) (int64, error) {
	query, err := sqlparser.ParseAndBind(`SELECT rows_copied FROM _vt.vreplication WHERE id=%a`,
		sqltypes.Int32BindVariable(id),
	)
	if err != nil {
		return 0, err
	}
	r, err := vr.dbClient.Execute(query)
	if err != nil {
		return 0, err
	}
	if len(r.Rows) != 1 || len(r.Rows[0]) != 1 {
		return 0, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "did not get expected single row value when getting rows_copied for workflow id: %d", id)
	}
	return r.Rows[0][0].ToInt64()
}
```

Reply:
@mattlord @olyazavr this code:
```go
	row := r.Named().Row()
	if row == nil {
           ...
        }
```
is idiomatic by design, meaning `Row()` will return `nil` when the number of rows != 1.
With the above code it's OK to both skip `if len(r.Rows) == 0` as well as `if len(r.Rows) != 1` and I'd say it's a handy shortcut for single-row responses (this is what `.Row()` was designed for).

Reply:
> In the latter you're doing the length check twice

I was actually pointing out that `if len(r.Rows) == 0` was not needed; so the code would be reduced to:

```go
        row := r.Named().Row()
	if row == nil {
		return 0, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "Cannot find unique workflow for UUID: %+v", uid)
	}
	return row.AsInt64("rows_copied", 0), nil
```

But it's not a big deal. I just saw a use of `Row()` which could be reduced. No strong feelings, and it's as you say a matter of preference.

Reply:
@shlomi-noach ah, OK. So we're mostly in agreement then. I think that your snippet is nicer to look at (more of an OOP design), but it still has one remaining issue, the third one I noted above:
  - If we for some reason got an unexpected/corrupted value back in the result, rather than knowing that because the type conversion failed, we again get back a 0 w/o knowing anything bad happened because AsInt64() will just return the default

That should be exceedingly rare though so probably not a practical issue. 

@olyazavr hopefully all of this made some sense! 😆 Whichever you prefer is fine with me:
```
	r, err := vr.dbClient.Execute(query)
	if err != nil {
		return 0, err
	}
	if len(r.Rows) != 1 {
		return 0, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "did not get expected single row value when getting rows_copied for workflow id: %d", uid)
	}
	return r.Rows[0][0].ToInt64()
```

```
	r, err := vr.dbClient.Execute(query)
	if err != nil {
		return 0, err
	}
        row := r.Named().Row()
	if row == nil {
		return 0, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "Cannot find unique workflow for id: %d", uid)
	}
	return row.AsInt64("rows_copied", 0), nil
```

I don't disagree with @shlomi-noach that the latter is nicer/more elegant as a reader IMO. It's more costly -- w/o much value in this case IMO -- but it's minimal. Just a reminder that either one of these is slightly more precise than the original, and the log message is corrected (it's not a UUID).

Sorry for all the noise! Again, happy to chat directly if it's easier.

Reply:
Either way looks fine to me.